Estimate gsof header.stamp based on PostitionTime or INSSolution

zthorson opened the pull request:
### Description

This is an optional functionality improvment to add correct timestamp values to some ROS Gsof Messages that currently write 0,0 for the timestamp.  It adds an additional flag to enable the feature, which defaults to off to maintain compatability with other releases.

### Reason

In the current code, any Gsof ROS message that does not have a GPSTime field in the structure will not write a timestamp value to the header.   In these cases the timestamp is always written as 0,0 in the GSOF message header.   If you attempt to use a TimeSynchronized message filter to group GSOF messages with NavFix (or INS) messages, this will cause them to fail since the GSOF message timestamp will never match.


### Method / Design

To address this a new parameter `estimate_gsof_time` was added to the ros parameters list.  This defaults to `false` in order to retain the same default behavior.

When `estimate_gsof_time` is set, the callback will use either the GSOF `PositionTime (#1)` gps time, or the `INSSolution (#49)` time as an estimate for other packets (Such as `Velocity (#8) or Attitude (#27)`).  It has a preference for using the INS message, as that use case is likely more common.

To get the best results, you would want the `PositionTime (#1)` or `INSSolution (#49)` message to have the highest update rate of all messages.  Ideally, it would also be a multiple of your lower frequency messages so that there is a good chance than GSOF will bundle the messages together on different pages of the same packet.  This would give the most accurate stamp.  Though in the worst case the timestamp would be off by 1/Hz s.

There exists one GSOF message `Attitude (#27)` that contains a special case.  This packet has a gps_time_ms field with no associated week.  To get the best timestamp, this packet will only use the GPSWeek from the `PositionTime (#1)` or `INSSolution (#49)` packet and use it's own gps_time_ms.  There exists a small chance of some odd behavior during the gps week rollover, but if the packets are synced well this should not occur.

### Testing

**Test Setup:**
- `ros:humble-ros-base` Docker image
  - ROS2 Humble
  - Ubuntu 22.04
  - Foxglove Studio for Packet visualization
  - BX992-INS GPS Units x 3

** GPS Configuration **

![image](https://github.com/user-attachments/assets/0907b06c-35d5-4c27-aad6-808eae9dc5d5)

---

**estimate_gsof_time = false**

Packets with no valid GPS time have a heading timestamp of 0,0:
![image](https://github.com/user-attachments/assets/777b8029-9443-4c83-95ed-7acd1be1afc9)

---

**estimate_gsof_time = true**

Packets with no valid GPS time now estimate their time based on `PositionTime (#1)`
![image](https://github.com/user-attachments/assets/44038c92-1b2f-4852-85ec-a9002cb2be35)

andre-nguyen committed:
feat: create SECURITY.md for security policy

e-tools-mend-production[bot] committed:
Add .whitesource configuration file

andre-nguyen committed:
feat: add Mend (whitesource)

Configure Mend for GitHub Enterprise

zthorson committed:
Update composable component name to match new namespace

  - Namespace was changed from trimble_driver_ros to trmb_ros but component was not updated
  - Update CMake file to match new namespace
  - Update launch composable node to the same namespace

andre-trimble committed:
ci: Update ubuntu ARM github runner label

zthorson committed:
Estimate gsof header.stamp based on PostitionTime or INSSolution

  - Many GSOF messages had header timestamps of 0,0, which causes issues for message filters
  - Added a flag to optionally use the closest known header timestamp to the GSOF message
  - This should be quite accurate of GSOF #1 or #49 has the highest system update rate

andre-nguyen commented on trimble_driver_ros/include/trimble_driver_ros/conversions.h:
```diff
@@ -60,6 +60,17 @@ struct HasGsofGpsTime : std::false_type {};
+// A few messages have a gps_time_ms field and no week (AttitudeInfo)
+// This will select cases where we have a time of week, but no week number
+template <typename T, typename = void>
+struct HasGPSmsOnly : std::false_type {};
+
+template <typename T>
+struct HasGPSmsOnly<T, decltype(T::gps_time_ms)> : std::true_type {};
+
+template <typename T>
+struct HasGPSmsOnly<T, decltype(T::gps_week, T::gps_week_number)> : std::false_type {};
```
For style consistency with other code and to disambiguate two sequential acronyms/abbreviations please change it to `HasGpsMillisOnly`

zthorson replied:
Sounds good to me.  I'll get that change in next week.

zthorson replied:
No longer applies.

andre-nguyen commented on trimble_driver_ros/include/trimble_driver_ros/gsof_client_ros.h:
```diff
@@ -130,8 +131,76 @@ class GsofClientRos : public rclcpp::Node {
+    auto gsof_msg        = message.as<NativeMessageType>();
+
+    // If we use the most recent timestamp to associate to the message, do so now.
+    // Fo these messages, we can also use the time of week to get better accuracy
```
```suggestion
    // For these messages, we can also use the time of week to get better accuracy
```

zthorson replied:
No longer applies

zthorson committed:
Add missed initializer to constructor

andre-nguyen commented on trimble_driver_ros/include/trimble_driver_ros/gsof_client_ros.h:
```diff
@@ -130,8 +131,76 @@ class GsofClientRos : public rclcpp::Node {
+      }
+
+      publisher->publish(ros_msg);
+    } else {
+      publisher->publish(toRosMessage(gsof_msg));
+    }
```
Can we refactor this into something like

```c++
uint16_t maybeGetGpsWeek() {
  if (ins_solution_) {
    return ins_solution_->gps_time.week;
  } else if (position_time_info_) {
    return position_time_info_->gps_week;
  } else {
    return 0;
  }
}
...
auto ros_msg = toRosMessage(gsof_msg);
if (estimate_gsof_time_) {
  ros_msg.header.stamp = getRosTimestamp(trmb::gsof::GpsTime{maybeGetGpsWeek(), gsof_msg->gps_time_ms});
}
publisher->publish(ros_msg);
```
?

zthorson replied:
See latest commit

andre-nguyen commented on trimble_driver_ros/include/trimble_driver_ros/gsof_client_ros.h:
```diff
@@ -130,8 +131,76 @@ class GsofClientRos : public rclcpp::Node {
+      publisher->publish(ros_msg);
+    } else {
+      publisher->publish(toRosMessage(gsof_msg));
+    }
```
see small refactor comment above

zthorson replied:
See latest commit

andre-nguyen commented on README.md:
```diff
@@ -110,6 +110,7 @@ Data Collector" (DCOL) protocol.
 | `child_frame`                           | `string`       | When publishing a ROS TF, the node will use this parameter as the `child_frame` for the published transform.                                                                                                                                                                                                                                       ...
 | `publish_tf`                            | `bool`         | Enable publishing of standard ROS TF.                                                                                                                                                                                                                                                                                                              ...
 | `publish_gsof_msgs`                     | `bool`         | Enable publishing of the ROS version of GSOF messages. These messages have almost exactly the same format as those defined in the GSOF protocol.                                                                                                                                                                                                   ...
+| `estaimte_gsof_time`                    | `bool`         | This will use the nearest PositionTime (#1) or INSSolution (#49) message as the header timestamp for messages that do not have GPS time in the GSOF protocol.  To get the most accurate time stamp, one of these two messages should be broadast equal to the greater than the fastest GSOF update rate. If this is disabled, some GSOF header tim ...
```
If I am understanding the feature correctly, `estimate_gsof_time == true` will only make sense if `time_source == gps`? If that is correct, can we add that to the docs?

zthorson replied:
Right now it affects all timestamps (at least in my testing).  Even with the time source set to `now`, the GSOF packets do not set any timestamp for packets that do not have a `gps_time` field present.  I just verified this on my system as well.

Adding a note here to correct my typo in: `estaimte` that should be `estimate`.

andre-nguyen replied:
My expectation is that is time_source is `now` then the time stamp should just always be overridden with `rclcpp::Clock()::now()` which is not the case for certain messages.

But I now understand what you were saying in your initial PR description about maintaining this behavior.

andre-nguyen commented on README.md:
```diff
@@ -110,6 +110,7 @@ Data Collector" (DCOL) protocol.
 | `child_frame`                           | `string`       | When publishing a ROS TF, the node will use this parameter as the `child_frame` for the published transform.                                                                                                                                                                                                                                       ...
 | `publish_tf`                            | `bool`         | Enable publishing of standard ROS TF.                                                                                                                                                                                                                                                                                                              ...
 | `publish_gsof_msgs`                     | `bool`         | Enable publishing of the ROS version of GSOF messages. These messages have almost exactly the same format as those defined in the GSOF protocol.                                                                                                                                                                                                   ...
+| `estaimte_gsof_time`                    | `bool`         | This will use the nearest PositionTime (#1) or INSSolution (#49) message as the header timestamp for messages that do not have GPS time in the GSOF protocol.  To get the most accurate time stamp, one of these two messages should be broadast equal to the greater than the fastest GSOF update rate. If this is disabled, some GSOF header tim ...
```
```suggestion
| `estimate_gsof_time`                    | `bool`         | This will use the nearest PositionTime (#1) or INSSolution (#49) message as the ROS `std_msgs::msg::Header` timestamp for messages that do not have GPS time in the GSOF protocol.  To get the most accurate time stamp, one of these two messages should be broadcast at equal or greater than the highest GSOF update rate. If this is disabled, some GSOF translated messages will have their `std_msgs::msg::Header` timestamps at 0 sec, 0 ns.                     | 
```

zthorson replied:
Wording changes have been updated, thank you.

andre-nguyen replied:
Is it possible the wording change got lost in some commit?

zthorson replied:
Yeah, looks like something got messed up when I was switching between our production code and this PR branch.  It should be fixed now.

andre-nguyen commented on trimble_driver_ros/src/trimble_driver_ros/gsof_client_ros.cpp:
Can we add a warning in the constructor like if `estimate_gsof_time` is `true` and the time source is not GPS then `RCLCPP_WARN(..., "Setting estimate_gsof_time to true will be overridden by time_source set to \'now\'"` or something of the sort?

zthorson replied:
See above comments.  No time stamp is set even if the source is set to `now` for the code I have.

andre-nguyen replied:
I understand now and this sounds like a bug #10 
The time stamp does get set to `now` for gsof 49. So something is broken with the `now` operation mode.

andre-nguyen reviewed (changes requested):
Thank you for the feature, looks pretty important for products that don't output gsof 49.

I tried it out on my unit and something doesn't seem to work for me. If I set the rate of gsof 1 to 1Hz and gsof 27 to 5Hz. My attitude info timestamp doesn't follow the time of week.

![image](https://github.com/user-attachments/assets/e115509a-deaa-4bb2-9657-cd15e7b93d23)

I'm thinking there is a problem where the template isn't being resolved properly?

zthorson commented:
> I tried it out on my unit and something doesn't seem to work for me. If I set the rate of gsof 1 to 1Hz and gsof 27 to 5Hz. My attitude info timestamp doesn't follow the time of week.
> 
> I'm thinking there is a problem where the template isn't being resolved properly?

I think I spot a problem with the template, the message I was trying to target uses a different name for the field:
https://github.com/trimble-oss/trimble_driver_ros/blob/ae97136a55e3e6e12e763f4e992f1081dcce287c/trimble_driver_ros/include/trimble_driver/gsof/message.h#L768-L770

Since I'm really only trying to target Attitude Info (the only one that has a gps time in ms but no week info), I think it may make more sense to just write a specific template override for that particular message and avoid adding a special templating case.  I'll look into it.

andre-nguyen commented:
> Since I'm really only trying to target Attitude Info (the only one that has a gps time in ms but no week info), I think it may make more sense to just write a specific template override for that particular message and avoid adding a special templating case. I'll look into it.

I wouldn't be against renaming the field to `gps_time_ms`, but I can confirm that this is the only message that doesn't include the gps week. A template specialization would make sense.

zthorson committed:
Update template to handle special timestamp messages

  - Better handle AttitudeInfo case where gps time in ms is available
  - Handle BasePositionAndQualityIndicator, PositionTimeInfo, and CurrentTime correctly
    - Updating field names in these would reduce the number of specialized templates needed at the expense of matching to documented field names

zthorson commented:
> > Since I'm really only trying to target Attitude Info (the only one that has a gps time in ms but no week info), I think it may make more sense to just write a specific template override for that particular message and avoid adding a special templating case. I'll look into it.
> 
> I wouldn't be against renaming the field to `gps_time_ms`, but I can confirm that this is the only message that doesn't include the gps week. A template specialization would make sense.

There are a few others that have specialized names and out of order fields.   I ended up adding them to the specialized converter list as well, since I saw they were using the time estimate rather than their included time fields.  

I still have more work on this PR for cleanup, but feel free to take a look at the new template approach.  Otherwise I should have access to the hardware again late next week for more testing.

![image](https://github.com/user-attachments/assets/1542b0c8-793d-4214-a259-9b14176e5e6a)

zthorson committed:
Fixed typos and wording of the `estimate_gsof_time` parameter

  - Minor wording changes from PR

zthorson committed:
Updates made by clang-format

  - Ran: find . -iname '*.h' -o -iname '*.cpp' -o -iname '*.hpp' | xargs clang-format-14 -i -style=file --fallback-style=llvm
  - Excluded /util/wise_enum folder from results

andre-nguyen commented on trimble_driver_ros/include/trimble_driver_ros/gsof_client_ros.h:
```diff
@@ -140,14 +143,131 @@ class GsofClientRos : public rclcpp::Node {
+  }
+
+  /**
+   * Specialized callback for AttitudeInfo to handle a the case where only gps week in ms is available
```
```suggestion
   * Specialized callback for AttitudeInfo to handle the case where only gps week in ms is available
```

andre-nguyen commented on README.md:
```diff
@@ -110,6 +110,7 @@ Data Collector" (DCOL) protocol.
 | `child_frame`                           | `string`       | When publishing a ROS TF, the node will use this parameter as the `child_frame` for the published transform.                                                                                                                                                                                                                                       ...
 | `publish_tf`                            | `bool`         | Enable publishing of standard ROS TF.                                                                                                                                                                                                                                                                                                              ...
 | `publish_gsof_msgs`                     | `bool`         | Enable publishing of the ROS version of GSOF messages. These messages have almost exactly the same format as those defined in the GSOF protocol.                                                                                                                                                                                                   ...
+| `estimate_gsof_time`                    | `bool`         | This will use the nearest PositionTime (#1) or INSSolution (#49) message as the header timestamp for messages that do not have GPS time in the GSOF protocol.  To get the most accurate time stamp, one of these two messages should be broadast equal to the greater than the fastest GSOF update rate. If this is disabled, some GSOF header tim ...
```
```suggestion
| `estimate_gsof_time`                    | `bool`         | This will use the nearest PositionTime (#1) or INSSolution (#49) message as the header timestamp for messages that do not have GPS time in the GSOF protocol.  To get the most accurate time stamp, one of these two messages should be broadcast equal to or greater than the fastest GSOF update rate. If this is disabled, some GSOF header timestamps will be at 0 sec, 0 ns.                     | 
```

andre-nguyen commented on trimble_driver_ros/include/trimble_driver_ros/gsof_client_ros.h:
```diff
@@ -140,14 +143,131 @@ class GsofClientRos : public rclcpp::Node {
+                                 !HasSpecializedConv<NativeMessageType>::value,
+                             bool> = true>
+  void publishGsofMessage(const trmb::gsof::Message &message) {
+    auto publisher = getGsofPublisher<RosMessageType>(gsof_id);
+    auto gsof_msg  = message.as<NativeMessageType>();
+    auto ros_msg   = toRosMessage(gsof_msg);
+
+    // Check if we should estimate a header time stamp or leave it as default
+    if (estimate_gsof_time_) {
+      ros_msg.header.stamp = getRosTimestampEstimate();
+    }
+
+    publisher->publish(ros_msg);
```
I think I'm seeing that this code path creates an opportunity for `rclcpp::Clock::now()` to never be set. So for example with GSOF 2 and `time_source == now` we get time `0` which is unexpected.

On top of that, I believe this behavior was here in the main branch but for different reasons. The template [condition](https://github.com/trimble-oss/trimble_driver_ros/blob/ae97136a55e3e6e12e763f4e992f1081dcce287c/trimble_driver_ros/include/trimble_driver_ros/gsof_client_ros.h#L143) is incorrect.

That being said, as you mentioned in your initial PR description, time 0 is the current behavior and it is possible someone depends on it. We can do an API change in the future with proper version increment to indicate so.

Linking issue #10 for future reference.

zthorson replied:
Yeah, the behavior for this appears to be slightly undefined, as in every other case where GPS time isn't present now() is also broken.  I had been defaulting to letting it translate the message using the default rules in `conversions.h`.  Which ends up using GPS time if available, and nothing if it isn't.

For example:
https://github.com/trimble-oss/trimble_driver_ros/blob/ae97136a55e3e6e12e763f4e992f1081dcce287c/trimble_driver_ros/src/trimble_driver_ros/conversions.cpp#L353
https://github.com/trimble-oss/trimble_driver_ros/blob/ae97136a55e3e6e12e763f4e992f1081dcce287c/trimble_driver_ros/src/trimble_driver_ros/conversions.cpp#L608

If you want to support now() as a general override for all templates, then we would have to change the logic in the generic `HasStdMsgsHeader<RosMessageType>::value && !HasGsofGpsTime<NativeMessageType>::value && !HasSpecializedConv<NativeMessageType>::value` as well as the `HasStdMsgsHeader<RosMessageType>::value && HasSpecializedConv<NativeMessageType>::value && std::is_same<NativeMessageType, trmb::gsof::AttitudeInfo>::value` to:

```
    // Check if we should estimate a header time stamp or leave it as default
    if (estimate_gsof_time_) {
      ros_msg.header.stamp = getRosTimestampEstimate();
    } else {
      ros_msg.header.stamp = getRosTimestamp(0,0);
    }
```

I agree that this would change the current behavior of the system, which is why I originally did not put it in place.  

It does seem unlikely to be that someone would rely on a 0,0 timestamp.  I can imagine anyone using it may have put workarounds in place, but I can't think of how you could build a ROS2 system that required a zero timestamp.

andre-nguyen replied:
> It does seem unlikely to be that someone would rely on a 0,0 timestamp. I can imagine anyone using it may have put workarounds in place, but I can't think of how you could build a ROS2 system that required a zero timestamp.

Agreed

andre-nguyen reviewed (changes requested):
So last minor comments.

Thank you for your patience.

edit: Also tested against GSOF 1, 2, 16 but not 41 as I don't have a base station connection right now.

zthorson committed:
Apply suggestions from code review

- Fix typos in comments and readme

Co-authored-by: Andre Nguyen <andre-nguyen@users.noreply.github.com>

zthorson committed:
Update Readme.md wording for missed change.

andre-nguyen commented:
![image](https://github.com/user-attachments/assets/2df6f40d-0db6-41b4-b7e9-fca032c67023)

Can confirm GSOF 41 is functional on my end.

Have you had a chance to test this PR on your own hardware?
